src/core/qgsactionmanager.cpp

```diff
@@ -56,21 +57,68 @@ QUuid QgsActionManager::addAction( QgsAction::ActionType type, const QString &na
 
 void QgsActionManager::addAction( const QgsAction &action )
 {
+  QgsDebugMsg( "add action " + action.name() );
   mActions.append( action );
+  if ( mLayer && mLayer->dataProvider() && !action.notificationMessage().isEmpty() )
+  {
+    mLayer->dataProvider()->setListening( true );
+    if ( !mOnNotifyConnected )
+    {
+      QgsDebugMsg( "connecting to notify" );
+      connect( mLayer->dataProvider(), &QgsDataProvider::notify, this, &QgsActionManager::onNotifyRunActions );
+      mOnNotifyConnected = true;
+    }
+  }
+}
+
+void QgsActionManager::onNotifyRunActions( const QString &message )
+{
+  for ( const QgsAction &act : qgsAsConst( mActions ) )
+  {
+    if ( !act.notificationMessage().isEmpty() && act.notificationMessage() == message )
+    {
+      if ( !act.isValid() || !act.runable() )
+        continue;
+
+      QgsExpressionContext context = createExpressionContext();
+
+      Q_ASSERT( mLayer ); // if there is no layer, then where is the notification coming from ?
+      context << QgsExpressionContextUtils::layerScope( mLayer );
+
+      QString expandedAction = QgsExpression::replaceExpressionText( act.command(), &context );
+      if ( expandedAction.isEmpty() )
+        continue;
+      runAction( QgsAction( act.type(), act.name(), expandedAction, act.capture() ) );
+    }
+  }
 }
 
 void QgsActionManager::removeAction( const QUuid &actionId )
 {
   int i = 0;
-  Q_FOREACH ( const QgsAction &action, mActions )
+  for ( const QgsAction &action : qgsAsConst( mActions ) )
   {
     if ( action.id() == actionId )
     {
       mActions.removeAt( i );
```

Comment:
You definitely have to take a copy of `mActions` here or there be dragons.

Reply:
so, no `qgsAsConst( mActions )` if mActions is not const hugh ?

Reply:
Not if inside the loop `mActions.removeAt( i )` is called

Reply:
As far as I know `qgsAsConst` doesn't take a copy, does it @nyalldawson ?

Reply:
next line is a break, so I'm not sure it matters, or am I missing something ?

Reply:
but I would expect an undefined behavior if I were to take the next value (typical usecase for a reverse loop or something)

Reply:
Nope, you are right, sorry.

Reply:
pfffff, you had me frightened there ! But you also had me thinking, which is a good thing. Thanks.